File: sdk/csharp/libraries/microsoft.bot.solutions/Skills/SkillDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.TraceExtensions;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Solutions.Skills.Models;
using Microsoft.Extensions.Configuration;

namespace Microsoft.Bot.Solutions.Skills
{
    /// <summary>
    /// A sample dialog that can wrap remote calls to a skill.
    /// </summary>
    /// <remarks>
    /// The options parameter in <see cref="BeginDialogAsync"/> must be a <see cref="SkillDialogArgs"/> instance
    /// with the initial parameters for the dialog.
    /// </remarks>
    public class SkillDialog : Dialog
    {
        private readonly string _botId;
        private readonly ConversationState _conversationState;
        private readonly SkillHttpClient _skillClient;
        private readonly EnhancedBotFrameworkSkill _skill;
        private readonly Uri _skillHostEndpoint;

        public SkillDialog(ConversationState conversationState, SkillHttpClient skillClient, EnhancedBotFrameworkSkill skill, IConfiguration configuration, Uri skillHostEndpoint)
            : base(skill?.Id)
        {
            if (configuration == null)
            {
                throw new ArgumentNullException(nameof(configuration));
            }

            _botId = configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;
            if (string.IsNullOrWhiteSpace(_botId))
            {
                throw new ArgumentException($"{MicrosoftAppCredentials.MicrosoftAppIdKey} is not in configuration");
            }

            _skillHostEndpoint = skillHostEndpoint;
            _skillClient = skillClient ?? throw new ArgumentNullException(nameof(skillClient));
            _skill = skill ?? throw new ArgumentNullException(nameof(skill));
            _conversationState = conversationState ?? throw new ArgumentNullException(nameof(conversationState));
        }

        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default)
        {
            if (!(options is SkillDialogArgs dialogArgs))
            {
                throw new ArgumentNullException(nameof(options), $"Unable to cast {nameof(options)} to {nameof(SkillDialogArgs)}");
            }

            var skillId = dialogArgs.SkillId;

            await dc.Context.TraceActivityAsync($"{GetType().Name}.BeginDialogAsync()", label: $"Using activity of type: {dialogArgs.ActivityType}", cancellationToken: cancellationToken).ConfigureAwait(false);

            Activity skillActivity;
            switch (dialogArgs.ActivityType)
            {
                case ActivityTypes.Event:
                    var eventActivity = Activity.CreateEventActivity();
                    eventActivity.Name = dialogArgs.Name;
                    eventActivity.ApplyConversationReference(dc.Context.Activity.GetConversationReference(), true);
                    skillActivity = (Activity)eventActivity;
                    break;

                case ActivityTypes.Message:
                    var messageActivity = Activity.CreateMessageActivity();
                    messageActivity.Text = dc.Context.Activity.Text;
                    skillActivity = (Activity)messageActivity;
                    break;

                default:
                    throw new ArgumentException($"Invalid activity type in {dialogArgs.ActivityType} in {nameof(SkillDialogArgs)}");
            }

            ApplyParentActivityProperties(dc.Context, skillActivity, dialogArgs);
            return await this.SendToSkillAsync(dc, skillActivity, cancellationToken).ConfigureAwait(false);
        }

        public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default)
        {
            await dc.Context.TraceActivityAsync($"{GetType().Name}.ContinueDialogAsync()", label: $"ActivityType: {dc.Context.Activity.Type}", cancellationToken: cancellationToken).ConfigureAwait(false);

            if (dc.Context.Activity.Type == ActivityTypes.EndOfConversation)
            {
                await dc.Context.TraceActivityAsync($"{GetType().Name}.ContinueDialogAsync()", label: $"Got EndOfConversation", cancellationToken: cancellationToken).ConfigureAwait(false);
                return await dc.EndDialogAsync(dc.Context.Activity.Value, cancellationToken).ConfigureAwait(false);
            }

            // Just forward to the remote skill
            return await SendToSkillAsync(dc, dc.Context.Activity, cancellationToken).ConfigureAwait(false);
        }

        public override Task<DialogTurnResult> ResumeDialogAsync(DialogContext dc, DialogReason reason, object result = null, CancellationToken cancellationToken = default)
        {
            return Task.FromResult(EndOfTurn);
        }

        public override async Task EndDialogAsync(ITurnContext turnContext, DialogInstance instance, DialogReason reason, CancellationToken cancellationToken = default)
        {
            if (reason == DialogReason.CancelCalled || reason == DialogReason.ReplaceCalled)
            {
                await turnContext.TraceActivityAsync($"{GetType().Name}.EndDialogAsync()", label: $"ActivityType: {turnContext.Activity.Type}", cancellationToken: cancellationToken).ConfigureAwait(false);

                var activity = (Activity)Activity.CreateEndOfConversationActivity();
                ApplyParentActivityProperties(turnContext, activity, null);

                await SendToSkillAsync(null, (Activity)activity, cancellationToken).ConfigureAwait(false);
            }

            await base.EndDialogAsync(turnContext, instance, reason, cancellationToken).ConfigureAwait(false);
        }

        private static void ApplyParentActivityProperties(ITurnContext turnContext, Activity skillActivity, SkillDialogArgs dialogArgs)
        {
            // Apply conversation reference and common properties from incoming activity before sending.
            skillActivity.ApplyConversationReference(turnContext.Activity.GetConversationReference(), true);
            skillActivity.ChannelData = turnContext.Activity.ChannelData;
            skillActivity.Properties = turnContext.Activity.Properties;

            if (dialogArgs != null)
            {
                skillActivity.Value = dialogArgs?.Value;
            }
        }

        private async Task<DialogTurnResult> SendToSkillAsync(DialogContext dc, Activity activity, CancellationToken cancellationToken)
        {
            if (dc != null)
            {
                // Always save state before forwarding
                // (the dialog stack won't get updated with the skillDialog and things won't work if you don't)
                await _conversationState.SaveChangesAsync(dc.Context, true, cancellationToken).ConfigureAwait(false);
            }

            var response = await _skillClient.PostActivityAsync(_botId, _skill, this._skillHostEndpoint, activity, cancellationToken).ConfigureAwait(false);
            if (!(response.Status >= 200 && response.Status <= 299))
            {
                throw new HttpRequestException($"Error invoking the skill id: \"{_skill.Id}\" at \"{_skill.SkillEndpoint}\" (status is {response.Status}). \r\n {response.Body}");
            }

            return EndOfTurn;
        }
    }
}

File: sdk/typescript/libraries/bot-solutions/src/skills/skillDialog.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import {
    ActivityTypes,
    TurnContext, 
    ConversationState,
    SkillHttpClient} from 'botbuilder';
import {
    DialogContext,
    DialogInstance,
    DialogReason,
    DialogTurnResult, 
    Dialog} from 'botbuilder-dialogs';
import { IEnhancedBotFrameworkSkill } from './models/enhancedBotFrameworkSkill';
import { SkillDialogArgs } from './skillDialogArgs';
import { IBotSettingsBase } from '../botSettings';
import { Activity, ConversationReference } from 'botframework-schema';
import { ActivityEx } from '../extensions';


/**
 * A sample dialog that can wrap remote calls to a skill.
 * @remarks The options parameter in BeginDialogAsync must be a SkillDialogArgs instance with the initial parameters for the dialog.
 */
export class SkillDialog extends Dialog {
    private readonly botId: string; 
    private readonly conversationState: ConversationState;
    private readonly skillClient: SkillHttpClient;
    private readonly skill: IEnhancedBotFrameworkSkill;
    private readonly skillHostEndpoint: string;

    public constructor(
        conversationState: ConversationState,
        skillClient: SkillHttpClient,
        skill: IEnhancedBotFrameworkSkill,
        configuration: IBotSettingsBase,
        skillHostEndpoint: string
    ) {
        super(skill.id);
        if (configuration === undefined) { throw new Error ('configuration has no value'); }
        if (configuration.microsoftAppId === undefined || configuration.microsoftAppId === '') { throw new Error ('The bot ID is not in configuration'); }
        if (skillClient === undefined) { throw new Error ('skillClient has no value'); }
        if (skill === undefined) { throw new Error ('skill has no value'); }
        if (conversationState === undefined) { throw new Error ('conversationState has no value'); }
        
        this.botId = configuration.microsoftAppId;
        this.skillHostEndpoint = skillHostEndpoint;
        this.skillClient = skillClient;
        this.skill = skill;
        this.conversationState = conversationState;
    }

    /**
     * When a SkillDialog is started, a skillBegin event is sent which firstly indicates the Skill is being invoked in Skill mode,
     * also slots are also provided where the information exists in the parent Bot.
     * @param dc inner dialog context.
     * @param options options
     * @returns dialog turn result.
     */
    public async beginDialog(dc: DialogContext, options?: object): Promise<DialogTurnResult> {
        if (!(options instanceof SkillDialogArgs)) {
            throw new Error('Unable to cast \'options\' to SkillDialogArgs');
        }
        
        const dialogArgs: SkillDialogArgs = options;
        //let skillId = dialogArgs.skillId; //skillId is not being used, but for parity with C#, this line is commented instead of removed
        await dc.context.sendTraceActivity(`${ SkillDialog.name }.onBeginDialog()`, undefined, undefined, `Using activity of type: ${ dialogArgs.activityType }`);
        
        let skillActivity: Activity;

        switch (dialogArgs.activityType) {
            case ActivityTypes.Event:
                let eventActivity = ActivityEx.createEventActivity();
                eventActivity.name = dialogArgs.name;
                const reference: Partial<ConversationReference> = TurnContext.getConversationReference(dc.context.activity);
                eventActivity = ActivityEx.applyConversationReference(eventActivity, reference, true);
                skillActivity = eventActivity as Activity;
                break;
            case ActivityTypes.Message:
                const messageActivity = ActivityEx.createMessageActivity();
                messageActivity.text = dc.context.activity.text;
                skillActivity = messageActivity as Activity;
                break;
            default:
                throw new Error(`Invalid activity type in ${ dialogArgs.activityType } in ${ SkillDialogArgs.name }`);
        }
        
        this.applyParentActivityProperties(dc.context, skillActivity, dialogArgs);
        return await this.sendToSkill(dc, skillActivity);
    }

    /**
     * All subsequent messages are forwarded on to the skill.
     * @param innerDC Inner Dialog Context.
     * @returns DialogTurnResult.
     */
    public async continueDialog(dc: DialogContext): Promise<DialogTurnResult> {
        await dc.context.sendTraceActivity(`${ SkillDialog.name }.continueDialog()`, undefined, undefined, `ActivityType: ${ dc.context.activity.type }`);
        
        if (dc.context.activity.type === ActivityTypes.EndOfConversation)
        {
            await dc.context.sendTraceActivity(`${ SkillDialog.name }.continueDialog()`, undefined, undefined, 'Got EndOfConversation');
            return await dc.endDialog(dc.context.activity.value);
        }

        // Just forward to the remote skill
        return await this.sendToSkill(dc, dc.context.activity);
    }

    public async resumeDialog(dc: DialogContext, reason: DialogReason, result: Object): Promise<DialogTurnResult> {
        return SkillDialog.EndOfTurn;
    }

    public async endDialog(turnContext: TurnContext, instance: DialogInstance, reason: DialogReason): Promise<void> {
        if (reason === DialogReason.cancelCalled || reason === DialogReason.replaceCalled) {
            await turnContext.sendTraceActivity(`${ SkillDialog.name }.endDialog()`, undefined, undefined, `ActivityType: ${ turnContext.activity.type }`);

            const activity: Activity = ActivityEx.createEndOfConversationActivity() as Activity;
            this.applyParentActivityProperties(turnContext, activity);

            await this.sendToSkill(undefined, activity);
        }

        await super.endDialog(turnContext, instance, reason);
    }

    private applyParentActivityProperties(turnContext: TurnContext, skillActivity: Activity, dialogArgs?: SkillDialogArgs): void {
        // Apply conversation reference and common properties from incoming activity before sending.
        const reference: Partial<ConversationReference> = TurnContext.getConversationReference(turnContext.activity);
        skillActivity = ActivityEx.applyConversationReference(skillActivity, reference, true) as Activity;
        skillActivity.channelData = turnContext.activity.channelData;
        // PENDING, the property 'Properties' does not exists in Activity
        //skillActivity.properties = turnContext.activity.properties;

        if (dialogArgs !== undefined)
        {
            skillActivity.value = dialogArgs.value;
        }
    }

    private async sendToSkill(dc: DialogContext | undefined, activity: Activity): Promise<DialogTurnResult> {
        if (dc !== undefined)
        {
            // Always save state before forwarding
            // (the dialog stack won't get updated with the skillDialog and things won't work if you don't)
            await this.conversationState.saveChanges(dc.context, true);
        }

        const response = await this.skillClient.postToSkill(this.botId, this.skill, this.skillHostEndpoint, activity);
        if (!(response.status >= 200 && response.status <= 299))
        {
            throw new Error (`Error invoking the skill id: "${ this.skill.id }" at "${ this.skill.skillEndpoint }" (status is ${ response.status }).\r\n${ response.body }`);
        }

        return SkillDialog.EndOfTurn;
    }
}
